streaming/store.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package streaming
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/constants"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/internal"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/logger"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/profiles"
+)
+
+// This file contains functions specific to storing state for a minimal downtime migration.
+
+// Struct to store minimal downtime migration specific metadata in the JobMetadata field in the JobDetails table
+type MinimalDowntimeJobMetadata struct {
+	IsSharded bool
+}
+
+// PersistJobDetails stores all the metadata associated with a job orchestration for a minimal downtime migration in the metadata db. An example of this metadata is job level data such as the spanner database name.
+func PersistJobDetails(ctx context.Context, targetProfile profiles.TargetProfile, sourceProfile profiles.SourceProfile, conv *internal.Conv, migrationJobId string, isSharded bool) (err error) {
+	project, instance, dbName, err := targetProfile.GetResourceIds(ctx, time.Now(), sourceProfile.Driver, nil)
+	if err != nil {
+		err = fmt.Errorf("can't get resource ids: %v", err)
+		return err
+	}
+	dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, constants.METADATA_DB)
+	client, err := utils.GetClient(ctx, dbURI)
+	if err != nil {
+		err = fmt.Errorf("can't create client for db %s: %v", dbURI, err)
+		return err
+	}
+	defer client.Close()
+	if err != nil {
+		err = fmt.Errorf("can't create database client: %v", err)
+		return err
+	}
+	err = writeJobDetails(ctx, migrationJobId, isSharded, conv, dbName, time.Now(), client)
+	if err != nil {
+		err = fmt.Errorf("can't store generated resources for datashard: %v", err)
+		return err
+	}
+	logger.Log.Info(fmt.Sprintf("Generated resources stored successfully for migration jobId: %s. You can also look at the 'spannermigrationtool_metadata' database in your spanner instance to get this jobId at a later point of time.\n", migrationJobId))
```

Comment:
nit: Do we need \n while using Log.Info - I think it automatically adds a new line anyway

Reply:
It only does so when - 

```
Log.Info("some text")
```

When we use `fmt.Sprintf` it prints whatever the `Sprintf` returns, to a `\n` is needed.